core/src/main/java/org/apache/calcite/rel/rules/JoinDeriveEquivalenceFilterRule.java

```diff
@@ -0,0 +1,380 @@
+ *  <blockquote><pre>{@code
+ *  LogicalJoin(condition=[=($1, $5)], joinType=[inner])
+ *   LogicalFilter(condition=[>($1, 20)])
+ *     LogicalTableScan(table=[[hr, emps]])
+ *   LogicalFilter(condition=[>($0, 20)])
+ *     LogicalTableScan(table=[[hr, depts]])
+ *  }</pre></blockquote>
+ *  <p>the query plan
+ *  <blockquote><pre>{@code
+ *  LogicalJoin(condition=[=($1, $5)], joinType=[inner])
+ *   LogicalFilter(condition=[SEARCH($1, Sarg[(10..30)])])
+ *     LogicalTableScan(table=[[hr, emps]])
+ *   LogicalFilter(condition=[SEARCH($0, Sarg[(20..40)])])
+ *     LogicalTableScan(table=[[hr, depts]])
+ *  }</pre></blockquote>
+ *  <p> will convert to
+ *  <blockquote><pre>{@code
+ *  LogicalJoin(condition=[=($1, $5)], joinType=[inner])
+ *   LogicalFilter(condition=[SEARCH($1, Sarg[(20..30)])])
+ *     LogicalTableScan(table=[[hr, emps]])
+ *   LogicalFilter(condition=[SEARCH($0, Sarg[(20..30)])])
+ *     LogicalTableScan(table=[[hr, depts]])
+ *  }</pre></blockquote>
+ *  <p>Currently, the rule has some limitations:
+ *  <p>1. only handle partial predicates (comparison), but this can be extended to
+ *      support more predicates such as 'LIKE', 'RLIKE' and 'SIMILAR' in the future.
+ *  <p>2. only support simple condition inference, such as: {$1 = $2, $2 = 10} =&gt; {$1 = 10},
+ *     can not handle complex condition inference, such as conditions with functions, like
+ *     {a = b, b = abs(c), c = 1} =&gt; {a = abs(1)}
+ *  <p>3. only support discomposed literal, for example
+ *     it can infer {$1 = $2, $1 = 10} =&gt; {$2 = 10}
+ *     it can not infer {$1 = $2, $1 = 10 + 10} =&gt; {$2 = 10 + 10}
+ */
+
+@Value.Enclosing
+public class JoinDeriveEquivalenceFilterRule
+    extends RelRule<JoinDeriveEquivalenceFilterRule.Config> implements TransformationRule {
+
+  public JoinDeriveEquivalenceFilterRule(Config config) {
+    super(config);
+  }
+
+  @Override public void onMatch(RelOptRuleCall call) {
+    final Filter filter = call.rel(0);
+    final Join join = call.rel(1);
+
+    final RexBuilder rexBuilder = join.getCluster().getRexBuilder();
+    final RexSimplify simplify =
+        new RexSimplify(rexBuilder, RelOptPredicateList.EMPTY, RexUtil.EXECUTOR);
+
+    final RexNode originalCondition =
+        prepare(rexBuilder, filter.getCondition(), join.getCondition());
+
+    final RexNode newCondition =
+        deriveEquivalenceCondition(simplify, rexBuilder, originalCondition);
+
+    if (arePredicatesEquivalent(rexBuilder, simplify, originalCondition, newCondition)) {
+      // if originalCondition and newCondition are equivalent, it means that the current
+      // Filter has either been derived or there is no room for derivation. if so,
+      // then we can stop.
+      return;
+    }
+
+    final Filter newFilter = filter.copy(filter.getTraitSet(), filter.getInput(), newCondition);
+
+    call.transformTo(newFilter);
+
+    // after derivation, the original filter can be pruned
+    call.getPlanner().prune(filter);
```

Comment:
I am pretty sure we do not want to do this.  If their is a large tree, then this could cause problems.

Reply:
ok, this might be too radical.

Reply:
Does this break anything if removed?  Why would it be to radical?